examples/batching_strategies/volume_batching/src/volume_batching.cc

```diff
@@ -0,0 +1,184 @@
+/// \param request The request to be added to the pending batch.
+/// \param userp The placeholder for backend to store and retrieve information
+/// about this pending batch. When the callback returns, this should reflect
+/// the latest batch information.
+/// \param should_include The pointer to be updated on whether the request was
+/// included in the batch.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONSERVER_Error*
+TRITONBACKEND_ModelBatchIncludeRequest(
+    TRITONBACKEND_Model* model, TRITONBACKEND_Request* request, void* userp,
+    bool* should_include)
+{
+  // Default should_include to false in case function returns error.
+  *should_include = false;
+
+  // Get current remaining batch volume.
+  unsigned int* remaining_volume = static_cast<unsigned int*>(userp);
+
+  // Get request's volume in bytes.
+  unsigned int pending_volume = 0;
+
+  uint32_t input_count;
+  auto err = TRITONBACKEND_RequestInputCount(request, &input_count);
+  if (err)
+    return err;
+
+  TRITONBACKEND_Input* input;
+  size_t data_byte_size;
+
+  for (size_t count = 0; count < input_count; count++) {
+    auto err =
+        TRITONBACKEND_RequestInputByIndex(request, count /* index */, &input);
+    if (err)
+      return err;
+    err = TRITONBACKEND_InputProperties(
+        input, nullptr, nullptr, nullptr, nullptr, &data_byte_size, nullptr);
+    if (err)
+      return err;
+    pending_volume += static_cast<unsigned int>(data_byte_size);
+  }
+
+  // Print remaining volume for debugging purposes.
+  std::cout << "Pending volume : " << pending_volume << std::endl;
+  std::cout << "Remaining volume : " << *remaining_volume << std::endl;
+
+  // Check if there is enough remaining volume for this request.
+  // If so, include this request. Otherwise, do not.
+  if (pending_volume <= *remaining_volume) {
+    *should_include = true;
+    *remaining_volume = *remaining_volume - pending_volume;
+  } else {
+    *should_include = false;
+  }
+
+  return nullptr;  // success
+}
+
+/// Callback to be invoked when Triton has begun forming a batch.
+/// \param model The backend model for which Triton is forming a batch.
+/// \param userp The placeholder for backend to store and retrieve information
+/// about this pending batch.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONSERVER_Error*
+TRITONBACKEND_ModelBatchInitialize(TRITONBACKEND_Model* model, void** userp)
+{
+  // Userp will point to an unsigned integer representing the remaining volume
+  // in bytes for this batch.
+
+  // Read the user-specified bytes from the model config.
+  TRITONSERVER_Message* config_message;
+  TRITONBACKEND_ModelConfig(model, 1 /* config_version */, &config_message);
+
+  const char* buffer;
+  size_t byte_size;
+
+  uint64_t max_volume_bytes = 0;
+  std::string max_volume_bytes_str;
+
+  auto err =
+      TRITONSERVER_MessageSerializeToJson(config_message, &buffer, &byte_size);
```

Comment:
This executes on every new batch. These calls are specific to each batch/payload. I'll have to think through this design a bit more. I'm not sure whether there's a better way to access this information or to store it.

We don't currently provide a mechanism for caching data outside of the user pointer. Right now, the user stores all data in the user poiner and manages its memory, but we expect the memory to be freed after the finalize function is called.

CC: @GuanLuo

Reply:
This actually executes on every batching call (which can happen more than once per batch). I thought about this some more, and one way we could modify the design to allow for caching is to add initialization and finalization functions for a cached variable. Something along the lines of: `TRITONBACKEND_ModelBatchInitializeCache(TRITONBACKEND_Model* model, void** userp_cached)` and `TRITONBACKEND_ModelBatchFinalizeCache(void* userp_cached)`. Then, add the userp_cached as an argument to `TRITONBACKEND_ModelBatchIncludeRequest.`

This means this feature has 5 new API calls rather than 3, but I think it'd be much cleaner for the user and more efficient. Let me know what you both think. @GuanLuo, would you be okay with these changes to the design?

Reply:
Even with the new API functions, we still have the problem of deserializing JSON for every batch when the values are known at the time of model configuration loading. While the searched path for the batching library is defaulted to `model_version->model->model_repository` the values which are read for processing batches are from the model config. Would we be able to store this in the Model itself? Maybe there is a better place to put this?

Reply:
I created an example implementation of the new API and put it in [this draft PR](https://github.com/triton-inference-server/backend/pull/76).

Essentially, we'd be calling the new API functions once each, so the user should do all one-time initialization then. In the cast of volume_batching.cc, we'd offload all the model config reading into those functions and save the max_volume_bytes. So the deserialization would only happen once, and yep, we'd store the user pointer in the model.